### ISSUE TITLE
Model with low prediction - need additional Operations that can be done during Training. - Missing

### ISSUE THREAD

**Issue by GuyYar1** [lmanov1/DL_DiabeticRetinopathyStagePrediction#7]
List in General:

**Can Be Done Without Training Images**
1. Gamma Correction: Adjust image brightness.
2. Filtering: Apply Gaussian blur, median filter, etc.
3. Color Space Transformation: Convert between color spaces like RGB to HSV.
4. Histogram Equalization: Enhance contrast by redistributing pixel intensities.
5. Normalization: Ensure images are normalized consistently.

**Requires Training Images**

1. Data Augmentation: Rotate, flip, scale, crop images, etc.
2. Use of Pre-trained Feature Extractors: Use models like VGG, ResNet as feature extractors.
3. Transfer Learning: Fine-tune pre-trained models on your dataset.
4. Ensemble Learning: Combine predictions from multiple models.
5. Regularization Techniques: Apply dropout, weight decay during training.

1. Data Augmentation
Data augmentation involves creating modified versions of images to expand your training dataset. This can help improve model robustness by exposing it to various transformations. Common techniques include:

Rotation: Rotate images to different angles.
Flipping: Horizontally or vertically flip images.
Scaling and Cropping: Randomly zoom in on images or crop them.
Color Jittering: Slightly adjust brightness, contrast, saturation, and hue.
Affine Transformations: Apply random translations, rotations, and scalings.
Libraries like torchvision provide simple functions for data augmentation:

python
Copy code
from torchvision import transforms

augmentation = transforms.Compose([
    transforms.RandomHorizontalFlip(),
    transforms.RandomRotation(10),
    transforms.ColorJitter(brightness=0.2, contrast=0.2, saturation=0.2),
])
2. Histogram Equalization
This technique improves the contrast of images by redistributing pixel intensity values, making the features more pronounced. It is particularly effective for enhancing images with poor contrast.

You can use libraries like OpenCV or PIL for this:

python
Copy code
import cv2

img = cv2.imread('path_to_image.jpg')
img_yuv = cv2.cvtColor(img, cv2.COLOR_BGR2YUV)
img_yuv[:,:,0] = cv2.equalizeHist(img_yuv[:,:,0])
img_output = cv2.cvtColor(img_yuv, cv2.COLOR_YUV2BGR)
3. Gamma Correction
Gamma correction adjusts the brightness of images by applying a power-law transformation. This helps to enhance or diminish the brightness in images, making features more visible.

python
Copy code
def adjust_gamma(image, gamma=1.0):
    invGamma = 1.0 / gamma
    table = [((i / 255.0) ** invGamma) * 255 for i in range(256)]
    return cv2.LUT(image, np.array(table).astype("uint8"))
4. Filtering
Applying filters can help smooth out images, remove noise, or highlight edges. Common filters include:

Gaussian Blur: Reduces image noise and detail.
Median Filter: Reduces noise while preserving edges.
Example of Gaussian Blur using OpenCV:

python
Copy code
blurred_image = cv2.GaussianBlur(image, (5, 5), 0)
5. Color Space Transformation
Changing the color space (e.g., from RGB to HSV or LAB) can help enhance certain features or make color-based segmentation easier.

python
Copy code
hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
6. Normalization
Ensure that images are normalized properly, not only using the mean and standard deviation of the dataset but also considering dynamic range adjustment. This ensures that pixel values are in a consistent range.

7. Use of Pre-trained Feature Extractors
Instead of PCA, consider using pre-trained networks (like VGG, ResNet, or EfficientNet) as feature extractors. Extract features from these models and then train a new model using these features.

python
Copy code
from torchvision import models

# Load pre-trained model
model = models.resnet50(pretrained=True)
model.eval()  # Set to evaluation mode
8. Transfer Learning
Fine-tuning a pre-trained model on your specific dataset can yield better results than training from scratch. Adjust only the final layers to match your classes while keeping the pre-trained weights for the earlier layers.

9. Ensemble Learning
Combine predictions from multiple models to improve accuracy. This can be achieved by averaging their outputs or using more sophisticated methods like stacking.

10. Regularization Techniques
Implement techniques like dropout or weight decay (L2 regularization) during training to prevent overfitting and improve generalization.

**Comment by GuyYar1** [comment]
![image](https://github.com/user-attachments/assets/83b3804f-8e6f-4256-ae73-c15deba45103)

**Comment by lmanov1** [comment]
We do quite a piece of all this. See in code.
The problem , as i understand this, with application of other technics (out
of those in your chatgpt friend' list we doesn't do)-  is that it requires
knowledge and intuition that dictate each specific choice , for the
specific reasons. Plus a solid and systematic methods for evaluation of a
change.
I luck such knowledge but can help with evaluation part.
.

**Comment by GuyYar1** [comment]
https://europepmc.org/article/pmc/pmc10001452#sec4-ijerph-20-04330
